### PR TITLE
Large JLSO files

### PR DESCRIPTION
**Pull request by rofinn**
Closing #21 

Pros:
- We can still use an existing file format like BSON for metadata
- We should be able to support petabyte scale files
- New format isn't really any less readable (objects were already just opaque serialized bytes)
- Changes are limited the on-disk format and the `read`/`write` functions. (e.g., v3 and v4 input/output is the same)

Cons:
- Logic for reading and writing is slightly uglier
- Might be some safety concerns with just having a continuous block of bytes that folks need to read after reading the BSON header?

### CODE REVIEW

**Comment by oxinabox** [comment]
Not reviewed properly, but this seems like a sensible way to store the data.

File: src/file_io.jl [hunk 1]
```diff
@@ -12,24 +11,43 @@ function Base.write(io::IO, jlso::JLSOFile)
...
+        "version" => string(jlso.version),
+        "julia" => string(jlso.julia),
+        "format" => string(jlso.format),
+        "compression" => string(jlso.compression),
+        "image" => jlso.image,
+        "project" => project_toml,
+        "manifest" => manifest_toml,
     )
+
+    if jlso.version < v"4"
```
Comment (oxinabox) [inline src/file_io.jl:25]: should we flip this so the main case of the current v4 is presented first?

File: src/file_io.jl [hunk 2]
```diff
@@ -107,3 +125,26 @@ end
 # FileIO Interface
 fileio_save(f, args...; kwargs...) = save(f.filename, args...; kwargs...)
 fileio_load(f, args...) = load(f.filename, args...)
+
+# Objects were initialially saved inline with the BSON document prior to v4
+# To support larger objects/files we started writing objects to the end of file and only save
+# names and sizes to the objects dict in the bson doc.
+# Since our ideal structure is one dict with object bytes we don't upgrade, but rather
+# extract both formats.
```
Comment (oxinabox) [inline src/file_io.jl:133]:
```suggestion
# Since our ideal structure is one dict with object bytes we don't solve this using the `upgrade`
# function but rather extract both formats here.
```
Can we also add a note in https://github.com/invenia/JLSO.jl/blob/master/src/upgrade.jl
saying that this code exist here, so that we know later when dropping support for v3 that we also need to delete this bit?

File: src/file_io.jl [hunk 3]
```diff
@@ -107,3 +125,26 @@ end
...
+function _extract_objects(io::IO, d)
```
Comment (oxinabox) [inline src/file_io.jl:134]: Should we make some note about the required state of the `io` object?
I.e. that pre-v4 it should be at end of file,
and for v4, it should have had the BSON metadata header already read off of it, and thus be at the start of the data section?

File: src/upgrade.jl [hunk 4]
```diff
@@ -124,6 +124,7 @@ function _upgrade_env(pkgs::Dict{String, VersionNumber})
     try
         mktempdir() do tmp
             Pkg.activate(tmp)
+            isempty(pkgs) && return _env()
```
Comment (oxinabox) [inline src/upgrade.jl:127]: why is this change here?

File: test/file_io.jl [hunk 5]
```diff
@@ -170,6 +170,35 @@ end
             JLSO.save("breakfast.jlso", "food" => "☕️🥓🍳", "time" => Time(9, 0)),
         )
     end
+    @testset "Size Limits" begin
+        # Tests that ensure we can work with objects larger than the BSON max
+        # doc/string/array size. Limited to size typemax(Int32).
+        # Instead of having a file size limited to 2.14 GB we should now support petabyte
```
Comment (oxinabox) [inline test/file_io.jl:176]:
```suggestion
        # Instead of having a file size limited to 2.14 GB we should support petabyte
```

File: test/file_io.jl [hunk 6]
```diff
@@ -170,6 +170,35 @@ end
...
+                obj = zeros(UInt8, typemax(Int32) + 1)
+                JLSO.save(
+                    joinpath(d, "large-object.jlso"),
+                    :X => obj;
+                    compression=:none,
+                )
+                loaded = JLSO.load(joinpath(d, "large-object.jlso"))
+                @test loaded == Dict(:X => obj)
+            end
+            @testset "Large Docs" begin
```
Comment (oxinabox) [inline test/file_io.jl:189]: I don't understand why this is called `large docs`

File: src/file_io.jl [hunk 7]
```diff
@@ -107,3 +125,26 @@ end
...
+    if VersionNumber(d["metadata"]["version"]) < v"4"
+        return Dict{Symbol, Vector{UInt8}}(Symbol(k) => v for (k, v) in d["objects"])
+    else
+        objects = Dict{Symbol, Vector{UInt8}}()
+
+        for (k, nb) in zip(d["object-names"], d["object-nbytes"])
+            bytes = read(io, nb)
+            if length(bytes) < nb
+                throw(ArgumentError("Failed to read all $nb bytes for object $k"))
```
Comment (oxinabox) [inline src/file_io.jl:143]: This is not an `ArgumentError`.
it is an `EOFError` though that doesn't have ability to attach a message.
OTOH I am not sure that that message is particularly useful, it might be or it might not be.
Do we exect this to occur because the file has been reandomly truncated (if so `EOFError` is fine)
or because some particular data will break the stream some how during righting (in which case we want to report the error)
We could strike a middle ground by logging something and using a `EOFError`.

(In general this kind of thing generally makes me thing julia (or any other programming language I've used) hasn't yet found the best way to do exceptions)

**Review by oxinabox — COMMENTED** [review]
Don't for get to add files to the specimens.
https://github.com/invenia/JLSO.jl/tree/master/test/specimens

My main concern would be how does a generic BSON reader know when to stop reading the BSON header?
Do we have a generic BSON reader to test what it does?
Maybe a python one?
The sensible thing to do is to stop when every opening `{` has been closed.
But i worry that some might just read the whole thing and then freakout about extra struff.

File: src/file_io.jl [hunk 8]
```diff
@@ -12,24 +11,43 @@ function Base.write(io::IO, jlso::JLSOFile)
...
+            io,
+            Dict{String, Union{Dict, Vector}}(
+                "metadata" => metadata,
+                "object-names" => string.(keys(jlso.objects)),
+                "object-nbytes" => [Int64(length(v)) for v in values(jlso.objects)]
+            )
+        )
+
+        # Now write our raw data bytes to the end of the IO
+        write(io, vcat(values(jlso.objects)...))
```
[the remark this answers is not in the document]
Comment (rofinn) [reply src/file_io.jl:49]: Yeah, that's fair. I was previously saving the bytes to a temporary array for debugging purposes, but that isn't necessary anymore.

File: src/upgrade.jl (hunk 4 above)
Comment (rofinn) [reply src/upgrade.jl:127]: Necessary to get tests to pass on julia 1.5 / nightly. I can move it to a separate pull request though.

**Review by oxinabox — APPROVED** [review]
This idea is all sound.
I can't think of any additional feedback I might need to give later.
Approved

File: test/file_io.jl (hunk 6 above)
Comment (rofinn) [reply test/file_io.jl:189]: Rather than having a single large vector (usually where the error hit) we want to test that the same issue occurs for large BSON docs. I suppose in BSON.jl the doc is called `BSONDict` so we could change it to that.

File: src/file_io.jl (hunk 7 above)
Comment (rofinn) [reply src/file_io.jl:143]: Yeah, unfortunately, this is trying to do both. I suppose the most likely error for writing issues is that the object ordering could get messed up, which this wouldn't catch anyways. I'll just change it to a `EOFError`.

**Comment by rofinn** [comment]
Cool, thanks for reviewing @oxinabox I'll make those changes and merge.

**Comment by rofinn** [comment]
> My main concern would be how does a generic BSON reader know when to stop reading the BSON header?

The BSON header has it's own EOF byte that should tell any reader to stop reading the doc at the end of the header. That assumes the reader is following the BSON spec though. I'll add some tests using the python bson package since that seems like the easiest options for julia interop.

File: src/upgrade.jl (hunk 4 above)
Comment (oxinabox) [reply src/upgrade.jl:127]: i have no idea what this is doing.

File: test/file_io.jl (hunk 6 above)
Comment (oxinabox) [reply test/file_io.jl:189]: or just large total size?

**Comment by oxinabox** [comment]
> I'll add some tests using the python bson package since that seems like the easiest options for julia interop.

They don't need to be CI tests.
Just doing a manual check once seems fine to me.
Adding CI tests about the interoperability of the BSON metadata  seems beyond the scope of this PR, but also reasonable to do one day.
(Low priority though)

File: test/file_io.jl (hunk 6 above)
Comment (rofinn) [reply test/file_io.jl:189]: Yeah, cause the max buffer length cannot exceed `typemax(Int32)` https://github.com/JuliaIO/BSON.jl/blob/master/src/write.jl#L32

Comment (oxinabox) [reply test/file_io.jl:189]:
```suggestion
            @testset "Large Total Size" begin
```

File: src/file_io.jl (hunk 3 above)
Comment (rofinn) [reply src/file_io.jl:134]: Hmmm, I wonder if it would be safer to include a `object-pos` array which specifies a position to seek to in the file io stream. In julia, if you seek to an invalid position and try to read then it'll just throw an `EOFError` anyways. This would also be helpful for libraries that don't support an IO based interface. For example, `bson.loads` correctly reads the header, but you're still stuck reading the first 4 BSON bytes to figure out the file offset to start reading the raw data from.

**Comment by rofinn** [comment]
> Just doing a manual check once seems fine to me.

Okay, yeah. I even have a little python snippet for extract raw object data from the file in python using the `bson` library. I've made a separate issue for that in #76

File: src/file_io.jl (hunk 3 above)
Comment (rofinn) [reply src/file_io.jl:150]: Opted to just include a positions array as well, so the io state isn't an externality to this function.

File: src/upgrade.jl (hunk 4 above)
Comment (rofinn) [reply src/upgrade.jl:127]: The `require_not_empty` was added to Pkg.jl for 1.5. This creates an early exit condition to work around that.

Comment (rofinn) [reply src/upgrade.jl:127]: https://github.com/JuliaLang/Pkg.jl/commit/87aab4a2b5e1e9213e4cde9f8049420a8e0e647f#diff-a43f827629f8ae12bbcbb218a293d325afcdda3a7f3c7016e2292840bf29ed1cR58

Comment (rofinn) [reply src/upgrade.jl:127]: This otherwise errors on the `Pkg.add` call in the `catch` below.
